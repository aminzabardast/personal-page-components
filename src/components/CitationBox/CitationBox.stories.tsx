import { Meta, StoryObj } from '@storybook/react'
import CitationBox from './CitationBox'

const meta: Meta<typeof CitationBox> = {
    title: 'CitationBox',
    component: CitationBox,
}

export default meta

type Story = StoryObj<typeof meta>

export const SingleLineWithCopy: Story = {
    args: {
        children: 'A',
        copyButton: true,
    },
}

export const MultiLineWithCopy: Story = {
    args: {
        children: ['A', 'B', 'C'],
        copyButton: true,
    },
}
