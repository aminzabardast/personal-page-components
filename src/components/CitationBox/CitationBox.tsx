import React, { ReactNode, Children } from 'react'
import { isArray, isNull, isUndefined } from 'lodash'
import { mdiContentCopy } from '@mdi/js'
import SvgIcon from '@mui/material/SvgIcon'
import Button from '@mui/material/Button'
import Box from '@mui/material/Box'

interface CitationBoxProps {
    children: ReactNode[] | ReactNode
    copyButton?: boolean
}

const textToHtml = (value: string) => {
    return (
        <Box
            component="pre"
            sx={{
                overflow: 'auto',
                padding: '10px',
                backgroundColor: (theme) =>
                    theme.palette.mode === 'dark' ? 'grey.800' : 'grey.300',
                borderColor: (theme) =>
                    theme.palette.mode === 'dark' ? 'grey.800' : 'grey.300',
                borderRadius: 2,
            }}
            dangerouslySetInnerHTML={{ __html: value }}
        ></Box>
    )
}

const CitationBox = ({ children, copyButton = false }: CitationBoxProps) => {
    let citation: string = ''
    if (!isNull(children) && !isUndefined(children)) {
        if (isArray(children)) {
            Children.forEach(children, (child) => {
                if (child) {
                    citation += `${child}\n`
                }
            })
        } else {
            citation = children as string
        }
    }
    const copyTextToClipboard = (value: string) => {
        return navigator.clipboard.writeText(value)
    }

    const copyFunction = () => {
        copyTextToClipboard(citation)
            .then(() => {})
            .catch(() => {})
    }
    return (
        <div data-testid="citation-box" style={{ position: 'relative' }}>
            {copyButton && (
                <Button
                    size="small"
                    onClick={copyFunction}
                    variant="contained"
                    endIcon={
                        <SvgIcon fontSize="inherit">
                            <path d={mdiContentCopy} />
                        </SvgIcon>
                    }
                    style={{
                        position: 'absolute',
                        top: 10,
                        right: 10,
                    }}
                >
                    Copy
                </Button>
            )}
            {textToHtml(citation)}
        </div>
    )
}

export default CitationBox
