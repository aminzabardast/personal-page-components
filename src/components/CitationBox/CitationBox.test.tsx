import { render, screen } from '@testing-library/react'
import CitationBox from './CitationBox'
import React from 'react'
import '@testing-library/jest-dom'

describe('<CitationBox>...</CitationBox>', () => {
    it('should render.', () => {
        render(<CitationBox>Cite!</CitationBox>)
        expect(screen.getByTestId('citation-box')).toBeInTheDocument()
    })
    it('should show the citation.', () => {
        render(<CitationBox>Cite!</CitationBox>)
        expect(screen.getByText('Cite!')).toBeInTheDocument()
    })
})
describe('<CitationBox copyButton>...</CitationBox>', () => {
    it('should have copy button.', () => {
        render(<CitationBox copyButton>Cite!</CitationBox>)
        expect(screen.getByRole('button')).toBeInTheDocument()
    })
    it('should copy the citation to clipboard when copy button is pressed.', () => {
        throw Error('Not Implemented.')
    })
    it('should change the button text to `Copied!` and then back to `Copy`.', () => {
        throw Error('Not Implemented.')
    })
    it('should copy the citations with line breaks.', () => {
        throw Error('Not Implemented.')
    })
})
describe('<CitationBox copyButton="false">...</CitationBox>', () => {
    it('should not show the copy button if it is disabled.', () => {
        throw Error('Not Implemented.')
    })
})
