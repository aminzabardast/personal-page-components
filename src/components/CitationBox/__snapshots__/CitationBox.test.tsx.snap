// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CitationBox> SnapShot Test should keep producing consistent result. 1`] = `
<div
  data-testid="citation-box"
  style={
    {
      "position": "relative",
    }
  }
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall css-6qwktu-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      {
        "position": "absolute",
        "right": 10,
        "top": 10,
      }
    }
    tabIndex={0}
    type="button"
  >
    Copy
    <span
      className="MuiButton-endIcon MuiButton-iconSizeSmall css-155nyw6-MuiButton-endIcon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"
        />
      </svg>
    </span>
  </button>
  <pre
    className="MuiBox-root css-1sod976"
    dangerouslySetInnerHTML={
      {
        "__html": "Line1! Line2!",
      }
    }
  />
</div>
`;
